toolchains/symbiflow.py

```diff
@@ -1001,6 +1003,451 @@ def check_env():
+                    'clocks':
+                        self.clocks,
+                    'environment_script':
+                        os.path.abspath('env.sh') + ' nextpnr fpga_interchange-' +
+                        self.device,
+                    'schema_dir':
+                        schema_dir,
+                    'output_format': #FIXME: remove when FASM support will be available
+                        'phys',
+                    'options':
+                        '--log nextpnr.log',
+                }
+
+                if self.fasm2bels:
+                    bitstream_device = None
+                    if self.device.startswith('a'):
+                        bitstream_device = 'artix7'
+                    if self.device.startswith('z010'):
+                        bitstream_device = 'zynq7'
+                    else:
+                        assert False, "Unsupported device type"
+
+                    part_json = os.path.join(
+                        self.dbroot, bitstream_device, self.family + self.part,
+                        'part.json'
+                    )
+                    symbiflow_options['yosys_additional_commands'] += [
+                        "plugin -i xdc",
+                        "yosys -import",
+                        "read_xdc -part_json {} {}".format(
+                            part_json, os.path.realpath(self.xdc)
+                        ),
+                        "clean",
+                        "write_blif -attr -param {}.eblif".format(
+                            self.project_name
+                        ),
+                    ]
+
+                self.edam = {
+                    'files': self.files,
+                    'name': self.project_name,
+                    'toplevel': self.top,
+                    'tool_options': {
+                        'symbiflow': symbiflow_options,
+                    }
+                }
+                self.backend = edalize.get_edatool('symbiflow')(
+                    edam=self.edam, work_root=self.out_dir
+                )
+                self.backend.configure("")
+
+            # TODO: change to .fasm when possible
+            self.backend.build_main(self.project_name + '.phys')
+            # self.run_steps()
+
+        self.add_runtimes()
+        self.add_wirelength()
+
+    def add_wirelength(self):
+        def get_wirelength(log_file):
+            wirelen = 0
+            with open(log_file, 'r') as fp:
+                for l in fp:
+                    l = l.strip()
+                    if 'wirelen =' in l:
+                        l_split = l.split()
+                        wirelen = l_split[l_split.index("wirelen") + 2]
+                        wirelen = ''.join(c for c in wirelen if c.isdigit())
+                        wirelen = int(wirelen)
+            return wirelen
+
+        route_log = os.path.join(self.out_dir, 'nextpnr.log')
+        self.wirelength = get_wirelength(route_log)
+
+    def max_freq(self):
```

Comment:
I might be wrong, but different nextpnr arch binaries should have the same frequency logs, therefore it should be possible to get this by inheritance.

Reply:
This will be common for both Xilinx and FPGAInterchange variants, but for now we non-generic [overridden method](https://github.com/SymbiFlow/fpga-tool-perf/pull/323/files#diff-69ab290f878020ed67b1008338e82df43674e16b1e45733d5f0d8e04b6601141R1081-R1108)